### ISSUE TITLE
discussion: do not include the primary key of max-one-row side in pk derivation of Inner(?) Join

### ISSUE THREAD

**Issue by BugenZhao** [risingwavelabs/risingwave#5764]
If we write an inner join whose one side is `SimpleAgg`, the primary key of output only contains the primary key of the other side since `SimpleAgg` has no primary key.

Things can actually be more general, as every max-one-row Plan Node can adapt this. This will benefit `DynamicFilter` which requires no output of the inner side.

**Comment by chenzl25** [comment]
For streaming query, max-one-row would always appear together with correlated sub-query which finally causes the aggregation becoming not `Simple` or more specifically with group by.

**Comment by chenzl25** [comment]
For the case of uncorrelated sub-query, it will produce a  nested-loop join which is unsupported right now.

**Comment by BugenZhao** [comment]
> For the case of uncorrelated sub-query, it will produce a nested-loop join which is unsupported right now.

I think some of them can be supported with [DynamicFilter](https://singularity-data.quip.com/AE06Ao1kAIaZ/RFC-Dynamic-Filter-A-New-Streaming-Operator)?

**Comment by chenzl25** [comment]
> > For the case of uncorrelated sub-query, it will produce a nested-loop join which is unsupported right now.
> 
> I think some of them can be supported with [DynamicFilter](https://singularity-data.quip.com/AE06Ao1kAIaZ/RFC-Dynamic-Filter-A-New-Streaming-Operator)?

Yes, you really remind me some of special nested-loop join can be transformed into dynamic filter. For dynamic filter, I think it's pk is the same with its left input, so we have already done?

**Comment by BugenZhao** [comment]
Yes. The `StreamDynamicFilter` is created when calling `LogicalJoin::to_stream` after we rewrite the plan to add primary keys, so even if we don't want the columns from the right side, its primary key is still added so the precondition of dynamic filter will be broken. That's one of the reasons for our only supporting simple agg as the inner side currently.

**Comment by chenzl25** [comment]
> Yes. The `StreamDynamicFilter` is created when calling `LogicalJoin::to_stream` after we rewrite the plan to add primary keys, so even if we don't want the columns from the right side, its primary key is still added so the precondition of dynamic filter will be broken. That's one of the reasons for our only supporting simple agg as the inner side currently.

Got it.

**Comment by BugenZhao** [comment]
I've had a try by hacking the pk derivation of Join side with `MaxOneRowVisitor`, it seems not to work since we'll have an intermediate state of `LogicalJoin` with two `Stream` plan nodes as the input. But the `MaxOneRowVisitor` is only implemented on logical plan nodes.

A trivial solution could be to repeat (or forward) the implementation to each stream plan node, but it sounds too verbose. Per some offline discussion with @st1page, the best solution might be #2279, after we finish the refactor of the optimizer.

**Comment by BugenZhao** [comment]
Actually this isn't just related to pk or stream key.

Since #12840, the stream key of global top-1 is already changed to an empty slice, but it's still not possible to use it as the inner side of dynamic filter. This is because, when performing `logical_rewrite_for_stream`, Join will always apply the column changes on inputs to its output, no matter whether the added column is indeed necessary.

https://github.com/risingwavelabs/risingwave/blob/76ecf8d81710b7fc1d6104baf54e6e549f94751d/src/frontend/src/optimizer/plan_node/logical_join.rs#L1632-L1644

So what we should fix is to avoid such unnecessary column from being inside the output of plan. This is what https://github.com/risingwavelabs/risingwave/pull/23634 is fixing.
